### DISCUSSION TITLE
getJobRetryUntil on Import Jobs

### DISCUSSION THREAD

**Discussion by ahinkle** [filamentphp/filament#13113] [General]
We encountered an issue where a bad import continuously retried for approximately a day. This led to significantly increased costs due to services like AWS SQS and exception logging. Why is this the default behavior? In Laravel, the default retry count is 3, which would limit retries to 3 per record. Why isn't this the standard setting instead?

via the documentation:

> 
> By default, the import system will retry a job for 24 hours. This is to allow for temporary issues, such as the database being unavailable, to be resolved. That functionality is defined in the getJobRetryUntil() method on the importer class:
> 
> use Carbon\CarbonInterface;
>  
> public function getJobRetryUntil(): ?CarbonInterface
> {
>     return now()->addDay();
> }
> 
> If you'd like to customize the retry time for jobs of a certain importer, you may override this method in your importer class. You can read more about job retries in the [Laravel docs](https://laravel.com/docs/queues#time-based-attempts).

https://filamentphp.com/docs/3.x/actions/prebuilt-actions/import#customizing-the-import-job-retries

> This is to allow for temporary issues, such as the database being unavailable, to be resolved.

I understand this but I find this to be an unlikely scenario vs. a bug in the code such as incorrect invalidation.

Otherwise love this importer! Thanks!

**Comment by spcbfr** [comment]
@ahinkle Hey, just came across this, I am curious, what was the bug that made the import job retry indefinitely? I also am using the importer and I got the same problem

**Reply by ahinkle** [reply]
Hey there @spcbfr,

It's not necessarily a bug, but the default Filament configuration is set to retry indefinitely for 24 hours when an import fails.

In my case, I encountered a simple validation issue where a field was being sent as nullable when it was not permitted. 
 
This was an easy fix, but I believe the retry limit should be set to use Laravel's default (which is 3 attempts) to prevent increased service costs and avoid unnecessary stress for developers.

**Reply by markdynamo6** [reply]
Where did you find the "default Filament configuration"? Can we also configure retries?

**Reply by ahinkle** [reply]
The `getJobRetryUntil` method; you would extend that.

**Reply by markdynamo6** [reply]
Cheers. Yes, we created a trait and used the following to only allow a single attempt:
```php
public function getJobRetryUntil(): ?CarbonInterface
{
    return now();
}
```

**Comment by sh4msi** [comment]
How did you apply $tries? I did not find a method for this
It looks like I need to rewrite the ExportCsv job?

@ahinkle

**Comment by mikailhamzat** [comment]
Any update on this?

**Comment by markdynamo6** [comment]
We had a similar issue with exports whereby a failed one that was perpetually retrying flooded our logs with errors. We were also using a load balancer and multiple hosts which unfortunately compounded the issue. We had set retries=3 but this was ignored. It appears that Export Job doesn't appear to use atomic locks or error throttling, hence the multi host problem. 

What's the recommended Filament approach to better control this?

**Comment by ahinkle** [comment]
Hi all, opened a PR to fix this here: https://github.com/filamentphp/filament/pull/14856

Open to feedback 👍
